src/root/rmem.d

```diff
@@ -126,42 +126,113 @@ else
+                stacktop = stackbottom;
+                stackleft = initialSize;
+            }
+
 
-    extern (C) void* allocmemory(size_t m_size) nothrow
+            memp = &stacktop;
+            memleft = &stackleft;
+      //      printf("BeginStackEnd - in Stack: %d - stackLeft: %d\n", memp != &heapp, stackleft);
+     //       assert(0);
+            return stacktop;
+        }
+
+        extern(C) void endStack(const void* stackBegin, ) nothrow
+        {
+            ptrdiff_t stacksize = (stacktop - stackBegin);
+            assert(stacksize > 0);
+            stackleft += stacksize;
+            stacktop = stacktop - stacksize;
+
+            if (stacktop == stackbottom) 
+            {
+                // if we discarded the last nesting stack
+                // switch back to heap
+                memp = &heapp;
+                memleft = &heapleft;
+            }
+        }
+    }
+    extern (C) void* allocmemory(const size_t _m_size) nothrow
     {
         // 16 byte alignment is better (and sometimes needed) for doubles
-        m_size = (m_size + 15) & ~15;
+        immutable m_size = (_m_size + 15) & ~15;
 
         // The layout of the code is selected so the most common case is straight through
-        if (m_size <= heapleft)
+  //      printf("StackMode : %d\n memLeft : %d\n", memp != &heapp, *memleft);
+		if (m_size <= *memleft)
         {
         L1:
-            heapleft -= m_size;
-            auto p = heapp;
-            heapp = cast(void*)(cast(char*)heapp + m_size);
+            *memleft -= m_size;
+            auto p = (*memp);
+            (*memp) = cast(void*)(cast(char*)p + m_size);
             return p;
         }
 
-        if (m_size > CHUNK_SIZE)
-        {
-            auto p = malloc(m_size);
-            if (p)
+        if (memp == &heapp) {
+            if (m_size > CHUNK_SIZE)
             {
-                return p;
+                auto p = malloc(m_size);
+                if (p)
+                {
+                    return p;
+                }
+                printf("Error: out of memory\n");
+                exit(EXIT_FAILURE);
             }
-            printf("Error: out of memory\n");
-            exit(EXIT_FAILURE);
-        }
 
-        heapleft = CHUNK_SIZE;
-        heapp = malloc(CHUNK_SIZE);
-        if (!heapp)
-        {
-            printf("Error: out of memory\n");
-            exit(EXIT_FAILURE);
-        }
+            *memleft = CHUNK_SIZE;
+           (*memp) = malloc(CHUNK_SIZE);
+            if (!(*memp))
+            {
+                printf("Error: out of memory\n");
+                exit(EXIT_FAILURE);
+            }
+        } else {
+            version (WithStack) {
+              //  stackbottom = realloc(stackbottom, stacksize + growBy);
+
+   /*             if (!stackbottom) {
+                    printf("Error: out of memory\n");
+                    exit(EXIT_FAILURE);
+                } */
+                stacktop = stackbottom + (CHUNK_SIZE / 2);
```

Comment:
We use the upper half of the stack as a circular buffer.
Since we should never have referenced that are used in the top part of the stack.